ZBUG-2334:- Added code to skip two factor auth check when auth request is coming from ZMPROV

madhavdhuppe opened the pull request:
**Problem:** 

- ZMPROV is broken when zimbraFeatureTwoFactorAuthRequired is TRUE
- Throwing nullPointerException when zmprov is called after zimbraFeatureTwoFactorAuthRequired is TRUE


**Fix:** 

- Added check while doing admin auth for account.getServer() null because getServer will be null for zmprov account as it is not a admin created account.
- we don't need to check 2fa for zmprov account.

**Testing to be done by QA:** 

1. Should not throw null pointer exception when the user does any zmprov operation after zimbraFeatureTwoFactorAuthRequired is TRUE
2. Check if auth is working as before.

**Note:**
The whitespaces are unavoidable as they occurred due to the tabSpace added when we create wrapper if loop.
Please use hide whitespace setting from GitHub while reviewing PR.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Zimbra/zm-mailbox?pullRequest=1167) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Zimbra/zm-mailbox?pullRequest=1167) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Zimbra/zm-mailbox?pullRequest=1167) it.</sub>

Prashantsurana commented on store/src/java/com/zimbra/cs/service/admin/Auth.java:
```diff
@@ -185,135 +185,139 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
-                    }
-                }
+                // Only perform 2fa authentication on admin account not on the resource accounts
+                if(acct != null && !Boolean.parseBoolean(acct.getAttr(AccountConstants.E_IS_SYSTEM_RESOURCE))) {
```
!acct.isIsSystemAccount()   -- should work

madhavdhuppe replied:
We don't have any method named isIsSystemAccount() in account class are you suggesting create one ?

Prashantsurana commented on store/src/java/com/zimbra/cs/service/admin/Auth.java:
```diff
@@ -185,135 +185,139 @@ public Element handle(Element request, Map<String, Object> context) throws Servi
-                }
+                // Only perform 2fa authentication on admin account not on the resource accounts
+                if(acct != null && !Boolean.parseBoolean(acct.getAttr(AccountConstants.E_IS_SYSTEM_RESOURCE))) {
+	                String deviceId = request.getAttribute(AccountConstants.E_DEVICE_ID, null);
```
I prefer to put entire block into private method

madhavdhuppe replied:
doing this

Prashantsurana commented on common/src/java/com/zimbra/common/soap/AccountConstants.java:
```diff
@@ -640,4 +640,5 @@
     public static final String E_SKIN_SELECTION_COLOR = "zimbraSkinSelectionColor";
     public static final String E_SKIN_FAVICON = "zimbraSkinFavicon";
     public static final String E_HOSTNAME = "hostname";
+    public static final String E_IS_SYSTEM_RESOURCE = "zimbraIsSystemResource";
```
if you use acct.isIsSystemAccount() this is not required.

Prashantsurana reviewed (approved):
Thanks

madhavdhuppe commented:
commit a37774079845adbdf1ad68d8e3c95ae4c3822b1d (HEAD -> develop, origin/develop, origin/HEAD)
Author: Madhav Dhuppe <44395134+madhavdhuppe@users.noreply.github.com>
Date:   Fri Jul 16 13:21:09 2021 +0530

    ZBUG-2334:- Added code to skip two factor auth check when auth request is coming from ZMPROV (#1167)

    * ZBUG-2334:- Added code to skip two factor auth check when auth request is coming from ZMPROV

    * ZBUG-2334: Moved 2FA entire code into private method.
